File: java/CombinationSum.java
/*
 * Filename: NextPermutation.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 20 Mar 2015 05:59:58 PM EDT
 */

public class Solution {
	public List<List<Integer>> combinationSum(int[] candidates, int target) {
		List<List<Integer>> ret = new ArrayList<List<Integer>>();
		List<Integer> tuple = new ArrayList<Integer>();

		if (candidates == null || candidates.length == 0) {
			return ret;
		}

		Arrays.sort(candidates);
		combinationSumAux(candidates, target, tuple, ret, 0);
		return ret;
	}

	private void combinationSumAux(int[] candidates, int target, List<Integer> tuple, List<List<Integer>> ret, int bound) {
		if (target < 0) {
			return;
		}

		if (target == 0) {
			ret.add(new ArrayList<Integer>(tuple));
			return;
		}

		for (int i = bound; i < candidates.length; i++) {
			tuple.add(candidates[i]);
			combinationSumAux(candidates, target - candidates[i], tuple, ret, i);
			tuple.remove(tuple.size() - 1);
		}
	}
}

File: cpp/combination_sum.cc
/*
 * Filename: combination_sum.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 25 Jul 2015 05:20:58 PM EDT
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		vector<vector<int> > combinationSum(vector<int>& candidates, int target) {
			vector<vector<int> > res;
			vector<int> v;
			sort(candidates.begin(), candidates.end());
			combination_sum_aux(candidates, 0, target, v, res);
			return res;
		}

	private:
		void combination_sum_aux
				(vector<int>& candidates, int i, int target, vector<int>& v, vector<vector<int> >& res) {
			if (target == 0) {
				res.push_back(v);
			} else if (target > 0) {
				for (vector<int>::size_type j = i; j < candidates.size(); j++) {
					v.push_back(candidates[j]);
					combination_sum_aux(candidates, j, target - candidates[j], v, res);
					v.pop_back();
				}
			}
		}
};
